src/falcon/falcon_c/rpo.c

```diff
@@ -0,0 +1,757 @@
+}
+
+void exp_acc(const uint64_t m, const uint64_t *base, const uint64_t *tail,
+             uint64_t *const res)
+{
+    for (uint64_t i = 0; i < m; i++)
+    {
+        for (uint64_t j = 0; j < STATE_WIDTH; j++)
+        {
+            if (i == 0)
+            {
+                *(res + j) = mult_mod_p(*(base + j), *(base + j));
+            }
+            else
+            {
+                *(res + j) = mult_mod_p(*(res + j), *(res + j));
+            }
+        }
+    }
+
+    for (uint64_t i = 0; i < STATE_WIDTH; i++)
+    {
+        *(res + i) = mult_mod_p(*(res + i), *(tail + i));
+    }
+}
+
+#include "rpo.h"
+
+/*
+ * Basic implementation of RPO128.
+ * Two TODOs remain:
+ * 1. Investigate the case of !FALCON_LE.
+ * 2. Investigate the padding rule in rpo128_inc_finalize as currently there are none.
+ */
+void rpo128_inc_init(inner_rpo128_context *rc)
+{
+    rc->dptr = 32;
+
+    /*
+     * Representation of an all-ones uint64_t is the same regardless
+     * of local endianness.
+     */
+    memset(rc->st.A, 0, sizeof rc->st.A);
+}
+
+void rpo128_inc_absorb(inner_rpo128_context *rc, const uint8_t *in, size_t len)
+{
+    size_t dptr;
+
+    dptr = (size_t)rc->dptr;
+    while (len > 0)
+    {
+        size_t clen, u;
+
+        /* 136 * 8 = 1088 bit for the rate portion in the case of SHAKE256
+            For RPO, this is 64 * 8 = 512 bits
+            The capacity for SHAKE256 is at the end while for RPO128 it is at the beginning */
+        clen = 96 - dptr;
+        if (clen > len)
+        {
+            clen = len;
+        }
+#if FALCON_LE // yyyLE+1
+        for (u = 0; u < clen; u++)
+        {
+            rc->st.dbuf[dptr + u] = in[u];
+        }
+#else  // yyyLE+0
+        for (u = 0; u < clen; u++)
+        {
+            rc->st.dbuf[dptr + u] = in[u];
+        }
+#endif // yyyLE-
+        dptr += clen;
+        in += clen;
+        len -= clen;
+        if (dptr == 96)
+        {
+            process_block(rc->st.A);
+            dptr = 32;
+        }
+    }
+    rc->dptr = dptr;
+}
+
+void rpo128_inc_finalize(inner_rpo128_context *rc)
+{
+    /*
+     * Set dptr to the end of the buffer, so that first call to
+     * extract will process the block.
+     */
+    rc->dptr = 96;
+}
+
+void rpo128_inc_squeeze(inner_rpo128_context *rc, uint8_t *out, size_t len)
+{
+    size_t dptr;
+
+    dptr = (size_t)rc->dptr;
+    while (len > 0)
+    {
+        size_t clen;
+
+        if (dptr == 96)
+        {
+            process_block(rc->st.A);
+            dptr = 32;
+        }
+        clen = 96 - dptr;
+        if (clen > len)
+        {
+            clen = len;
+        }
+        len -= clen;
+#if FALCON_LE // yyyLE+1
+        memcpy(out, rc->st.dbuf + dptr, clen);
+        dptr += clen;
+        out += clen;
+#else  // yyyLE+0
+        memcpy(out, rc->st.dbuf + dptr, clen);
+        dptr += clen;
+        out += clen;
+#endif // yyyLE-
+    }
+    rc->dptr = dptr;
+}
+
+void PQCLEAN_FALCON512_CLEAN_hash_to_point_rpo(
```

Comment:
Would it make sense to pull this function through the FFI to the Rust side and then have a test to make sure it works the same way as the Rust implementation of hash-to-point? This would also test C implementation of RPO.

Reply:
It is a bit cumbersome but it makes sense. I have done it as a test in `signature.rs`. Let me know if it should be moved somewhere else.